device.go

```diff
@@ -42,31 +47,110 @@ var (
 
 var deviceHandlerList = map[string]deviceHandler{
 	driverBlkType:  virtioBlkDeviceHandler,
 	driverSCSIType: virtioSCSIDeviceHandler,
 }
 
-func virtioBlkDeviceHandler(device pb.Device, spec *pb.Spec) error {
-	// First need to make sure the expected device shows up properly,
-	// and then we need to retrieve its device info (such as major and
-	// minor numbers), useful to update the device provided
-	// through the OCI specification.
-	devName := strings.TrimPrefix(device.VmPath, devPrefix)
-	checkUevent := func(uEv *uevent.Uevent) bool {
-		return (uEv.Action == "add" &&
-			filepath.Base(uEv.DevPath) == devName)
+func getDevicePCIAddress(pciID string) (string, error) {
+	tokens := strings.Split(pciID, "/")
+
+	if len(tokens) != 2 {
+		return "", fmt.Errorf("PCI Identifier for device should be of format [bridgeAddr/deviceAddr], got %s", pciID)
+	}
+
+	bridgeID := tokens[0]
+	deviceID := tokens[1]
+
+	// Deduce the complete bridge address based on the bridge address identifier passed
+	// and the fact that bridges are attached on the main bus with function 0.
+	pciBridgeAddr := fmt.Sprintf("0000:00:%s.0", bridgeID)
+
+	// Find out the bus exposed by bridge
+	bridgeBusPath := fmt.Sprintf(pciBusPathFormat, pciBridgeAddr)
+
+	files, err := ioutil.ReadDir(bridgeBusPath)
+	if err != nil {
+		return "", fmt.Errorf("Error with getting bridge pci bus : %s", err)
+	}
+
+	busNum := len(files)
+	if busNum != 1 {
+		return "", fmt.Errorf("Expected an entry for bus in %s, got %d entries instead", bridgeBusPath, busNum)
+	}
+
+	bus := files[0].Name()
+
+	// Device address is based on the bus of the bridge to which it is attached.
+	// We do not pass devices as multifunction, hence the trailing 0 in the address.
+	pciDeviceAddr := fmt.Sprintf("%s:%s.0", bus, deviceID)
+
+	bridgeDevicePCIAddr := fmt.Sprintf("%s/%s", pciBridgeAddr, pciDeviceAddr)
+	agentLog.WithField("completePCIAddr", bridgeDevicePCIAddr).Info("Fetched ")
+
+	return bridgeDevicePCIAddr, nil
+}
+
+func getBlockDeviceNodeName(s *sandbox, pciID string) (string, error) {
+	pciAddr, err := getDevicePCIAddress(pciID)
+	if err != nil {
+		return "", err
+	}
+
+	var devName string
+	var notifyChan chan string
+
+	fieldLogger := agentLog.WithField("pciID", pciID)
+
+	// Check if the PCI identifier is in PCI device map.
+	s.Lock()
+	for key, value := range s.pciDeviceMap {
+		if strings.Contains(key, pciAddr) {
+			devName = value
+			fieldLogger.Info("Device found in pci device map")
+			break
+		}
 	}
-	if err := waitForDevice(device.VmPath, devName, checkUevent); err != nil {
+
+	// If device is not found in the device map, hotplug event has not
+	// been received yet, create and add channel to the watchers map.
+	// The key of the watchers map is the device we are interested in.
+	// Note this is done inside the lock, not to miss any events from the
+	// global udev listener.
+	if devName == "" {
+		notifyChan := make(chan string)
```

Comment:
Probably better to create a buffered channel of 1 here. This way, the uevent listener won't get stuck waiting for someone to read the event.